cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/SsaInternalsCommon.qll

```diff
@@ -127,7 +127,79 @@ class AllocationInstruction extends CallInstruction {
   AllocationInstruction() { this.getStaticCallTarget() instanceof Cpp::AllocationFunction }
 }
 
+/**
+ * An abstract class for handling indirections.
+ *
+ * Extend this class to make a type behave as a pointer for the
+ * purposes of dataflow.
+ */
+abstract class Indirection extends Type {
+  /** Gets the type of this indirection. */
+  final Type getType() { result = super.getUnspecifiedType() }
+
+  /**
+   * Gets the number of indirections supported by this type.
+   *
+   * For example, the number of indirections of a variable `p` of type
+   * `int**` is `3` (i.e., `p`, `*p` and `**p`).
+   */
+  abstract int getNumberOfIndirections();
+
+  /**
+   * Holds if `deref` is an instruction that behaves as a `LoadInstruction`
+   * that loads the value computed by `addres`.
+   */
+  predicate isAdditionalDereference(Instruction deref, Operand address) { none() }
+
+  /**
+   * Holds if `value` is written to the address computed by `address`.
+   *
+   * `certain` is `true` if this write is guarenteed to write to the address.
+   */
+  predicate isAdditionalWrite(Node0Impl value, Operand address, boolean certain) { none() }
+
+  /**
+   * Gets the base type of this indirection.
+   *
+   * For example, the base type of `int*&` is `int*`, and the base type of `int*` is `int`.
+   */
+  abstract Type getBaseType();
+
+  /** Holds if there should be an additional taint step from `node1` to `node2`. */
+  predicate isAdditionalTaintStep(Node node1, Node node2) { none() }
+
+  /**
+   * Holds if the step from `opFrom` to `instrTo` should be considered a conversion
+   * from `opFrom` to `instrTo`.
+   */
+  predicate isAdditionalConversionFlow(Operand opFrom, Instruction instrTo) { none() }
+
+  /**
+   * Holds if writing the value `value` to an address with base `base` should
+   * be ignored.
+   */
+  predicate ignoreSourceVariableBase(BaseSourceVariableInstruction base, Node0Impl value) { none() }
+}
+
+private class PointerOrReferenceTypeIndirection extends Indirection, PointerOrReferenceType {
+  override int getNumberOfIndirections() { result = 1 + countIndirections(this.getBaseType()) }
+
+  override predicate isAdditionalDereference(Instruction deref, Operand address) { none() }
+
+  override predicate isAdditionalWrite(Node0Impl value, Operand address, boolean certain) { none() }
+
+  override Type getBaseType() { result = PointerOrReferenceType.super.getBaseType() }
+}
+
+predicate isDereference(Instruction deref, Operand address) {
```

Comment:
Yeah, there actually is. Since the `Indirection` class is defined by extending `Type` it means that every `isAdditionalDereference` that's implemented must ensure to also bind `this` to avoid duplicating the work across many `Indirection` classes. For example, let's suppose we implemented loading from a `LoadInstruction` by implementing the `isAdditionalDereference` class in `PointerOrReferenceTypeIndirection` like:
```ql
override predicate isAdditionalDereference(Instruction deref, Operand address) {
  deref.(LoadInstruction).getSourceAddressOperand() = address
}
```
this would definitely work, but since this doesn't bind `this` we'd have a cartesian product between all `PointerOrReferenceType`s and the columns in this predicate.

I _think_ we could fix this by implementing the predicate like this instead:
```ql
override predicate isAdditionalDereference(Instruction deref, Operand address) {
  address.getType() = this and // now we bind this so that every `Indirection` instance contributes a disjoint set of cases.
  deref.(LoadInstruction).getSourceAddressOperand() = address
}
```
but since we do actually want all `LoadInstruction` to be considered dereferences, I figured it was simplest to just implement this case outside the `Indirection` class.

A third option would be to avoid extending `Type`, and instead extend an IPA type. This IPA type would then have a branch for each `Indirection` we care about (the next PR will add an implementation of `Indirection` for iterators, for example), and we could implement `PointerOrReferenceTypeIndirection` by extending a unit branch of this IPA type to avoid the cartesian product between `PointerOrReferenceType` and the columns of `isAdditionalDereference`.

Reply:
I see the problem now.

It seems that the IPA solution would require us to extend the IPA type every time we want to add a type of indirection, right? If so, that makes it more difficult for users to add additional indirection types (in the unlikely case they start digging this deeply), which is not ideal.

So lets leave this as it is.